apollo-router/src/plugins/external.rs

```diff
@@ -320,69 +440,425 @@ impl Plugin for ExternalPlugin {
-            .buffered()
+            .buffer(20_000)
             .service(service)
             .boxed()
     }
 }
 
-type ExternalParams<'a> = (
-    Option<&'a HeaderMap<HeaderValue>>,
-    Option<serde_json::Value>,
-    Option<Context>,
-    Option<String>,
-);
+// -----------------------------------------------------------------------------------------
+async fn process_router_request_stage<C>(
+    http_client: C,
+    coprocessor_url: String,
+    sdl: Arc<String>,
+    mut request: router::Request,
+    request_config: RouterConf,
+) -> Result<ControlFlow<router::Response, router::Request>, BoxError>
+where
+    C: Service<hyper::Request<Body>, Response = hyper::Response<Body>, Error = BoxError>
+        + Clone
+        + Send
+        + Sync
+        + 'static,
+    <C as tower::Service<http::Request<hyper::Body>>>::Future: Send + 'static,
+{
+    // Call into our out of process processor with a body of our body
+    // First, extract the data we need from our request and prepare our
+    // external call. Use our configuration to figure out which data to send.
+    let (parts, body) = request.router_request.into_parts();
+    let bytes = hyper::body::to_bytes(body).await?;
+
+    let headers_to_send = request_config
+        .headers
+        .then(|| externalize_header_map(&parts.headers))
+        .transpose()?;
+
+    let body_to_send = request_config
+        .body
+        .then(|| serde_json::from_slice::<serde_json::Value>(&bytes))
+        .transpose()?;
+    let context_to_send = request_config.context.then(|| request.context.clone());
+    let sdl = request_config.sdl.then(|| sdl.clone().to_string());
+
+    let payload = Externalizable {
+        version: EXTERNALIZABLE_VERSION,
+        stage: PipelineStep::RouterRequest.to_string(),
+        control: Some(Control::default()),
+        id: TraceId::maybe_new().map(|id| id.to_string()),
+        headers: headers_to_send,
+        body: body_to_send,
+        context: context_to_send,
+        sdl,
+        uri: None,
+        service_name: None,
+    };
 
-fn prepare_external_params<'a>(
-    config: &'a BaseConf,
-    headers: &'a HeaderMap<HeaderValue>,
-    bytes: &'a Bytes,
-    context: &'a Context,
-    sdl: String,
-) -> Result<ExternalParams<'a>, BoxError> {
-    let mut headers_opt = None;
-    // Note: We have to specify the json type or it won't deserialize correctly...
-    // let mut payload_opt: Option<serde_json::Value> = None;
-    let mut payload_opt: Option<serde_json::Value> = None;
-    let mut context_opt = None;
-    let mut sdl_opt = None;
-
-    if config.body || config.headers {
-        if config.body {
-            payload_opt = Some(serde_json::from_slice(bytes)?);
-        }
-        if config.headers {
-            headers_opt = Some(headers);
-        }
+    tracing::debug!(?payload, "externalized output");
+    request.context.enter_active_request().await;
+    let co_processor_result = payload.call(http_client, &coprocessor_url).await;
```

Comment:
we should have tracing propagation to the external service (`TextMapPropagator` etc)